Add background image to mini/main console

Edru2 opened the pull request:
#### Brief overview of PR changes/additions
new function:
```lua
setConsoleBackgroundImage([name], imgPath, [mode]) -- name and mode are optional
```
Adds the ability to add an Image as background for a console.
The mode parameter gives you the ability to chose different background drawing modes as:
1 "border" (the background image is stretched) , 2 "center" (the background image is in the center), 3 "tile" (the background image is 'tiled'), 4 "style" (This mode is special and allows you to choose the background image style-sheet by yourself which means every possible qss background image style is settable)

to reset the background image:
```lua
resetConsoleBackgroundImage([name]) -- name is optional
```

decho and hecho allow to use transparency parameter for  the background to write on the console.
cecho has a new color transparent (only works for background) added.

Geyser.MiniConsole has a wrapper integrated
```lua
myminiconsole:setBackgroundImage(imgPath, [mode])
myminiconsole:resetBackgroundImage()
```

#### Motivation for adding to Mudlet
More possibilities to make a gui better looking :smiley: 

#### Other info (issues closed, discussion etc)
example code (creates a miniconsole with mudlet spashscreen as background and test decho and hecho)

```lua
miniconsole_withBgImg = Geyser.MiniConsole:new({name = "BackgroundonConsole", x=0, y=0, width = "20%", height = "40%"})
miniconsole_withBgImg:setBackgroundImage(":/Mudlet_splashscreen_development.png", "center")

for i = 1, 6 do
  miniconsole_withBgImg:decho("<:0,0,0,130> This is a decho test with semi-transparent background\n")
end

for i = 1, 6 do
  miniconsole_withBgImg:hecho("#,80000000 This is a hecho test with semi-transparent background\n")
end

for i = 1, 2 do
  miniconsole_withBgImg:decho("<:0,0,0> This is a decho test with no transparent background\n")
end

for i = 1, 2 do
  miniconsole_withBgImg:hecho("#,000000 This is a hecho test with no transparent background\n")
end
```

Edru2 committed:
first success

Edru2 committed:
add background image to miniconsole transparency for cecho,decho,hecho

Added ability to add background image to miniconsole.
cecho, hecho, decho support transparency now

Edru2 committed:
add image also to main console

+ use mpBackground for backgroundcolors
to increase performance

Edru2 committed:
Update TLuaInterpreter.cpp

vadi2 commented:
This is very cool! Can you add some benchmarks so we can measure it doesn't affect Mudlet's speed?

Edru2 committed:
give splitter palette to solve issue with darktheme

darktheme was able to change the background if the splitter had no color set.
this is fixed now

Edru2 committed:
still darktheme issue

Edru2 committed:
elements were hidden behind the background (fixed)

vadi2 commented on src/TTextEdit.cpp:
```diff
@@ -99,7 +99,7 @@ TTextEdit::TTextEdit(TConsole* pC, QWidget* pW, TBuffer* pB, Host* pH, bool isLo
     cursor.setShape(Qt::IBeamCursor);
     setCursor(cursor);
-    setAttribute(Qt::WA_OpaquePaintEvent); //was disabled
+    setAttribute(Qt::WA_OpaquePaintEvent, false); //was disabled
```
Comment no longer needed, right?

Edru2 commented:
I made some performance tests, but I don't know if they are really telling something.

For the tests I used once a the version downloaded here and the last ptb version for Linux.

But there seem to be no difference in speed in my echo test this PR performs even slightly faster which could be because in this PR the background doesn't need to be constantly redrawn (but I could be wrong about that).

**Benchmark-tests I used to copy and paste:**
simple server connect to localhost:10000
```python
#!/usr/bin/env python

import socket
import sys

TCP_IP = '127.0.0.1'
TCP_PORT = 10000
BUFFER_SIZE = 1024

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.bind((TCP_IP, TCP_PORT))
s.listen(1)

conn, addr = s.accept()
print ('Connection address:', addr)
while 1:
    data = conn.recv(BUFFER_SIZE)
    if not data: break
    print ("received data:", data)
    conn.send(data)  # echo
    send_data = b" "
    for i in range(0, 16):
        for j in range(0, 16):
            code = str(i * 16 + j)
            code2 = u"\u001b[38;5;" + code + "m " + code.ljust(4)
            send_data = send_data + bytes(code2,'utf8')
        conn.send(send_data)
        conn.send(bytes("\u001b[0m",'utf8'))
    conn.send(bytes("Finished!",'utf8'))
s.shutdown(1)
conn.close()
```

codes for tests in Mudlet
```lua 
function bench()
  createStopWatch("benchm.time")
  resetStopWatch("benchm.time")
  startStopWatch("benchm.time")
  
  if myTempTrigger then
    killTrigger(myTempTrigger)
    myTempTrigger = nil
  end
  
  myTempTrigger = myTempTrigger or tempTrigger("Finished!", 
    function() 
        local time = stopStopWatch("benchm.time")
        echo("\nResolved in: "..time.."\n")
        killTrigger(myTempTrigger)
        myTempTrigger = nil
        return
    end)
  
  send("Start BenchMark") 
end 

function bench()
  createStopWatch("benchm.time")
  resetStopWatch("benchm.time")
  startStopWatch("benchm.time")
  
  if myTempTrigger then
    killTrigger(myTempTrigger)
    myTempTrigger = nil
  end
  
  myTempTrigger = myTempTrigger or tempTrigger("Finished!", 
    function() 
        local time = stopStopWatch("benchm.time")
        echo("\nResolved in: "..time.."\n")
        killTrigger(myTempTrigger)
        myTempTrigger = nil
        return
    end)
  
  send("Start BenchMark") 
end 

function echoTest()
  createStopWatch("echo.time")
  resetStopWatch("echo.time")
  startStopWatch("echo.time")
  local i
  for i = 1, 50000 do
    cecho("<blue:yellow>THIS IS A TEST\n")
    cecho("<green:red>THIS IS A TEST THIS IS A TEST THIS IS A TEST THIS IS A TEST THIS IS A TEST THIS IS A TEST THIS IS A TEST THIS IS A TEST THIS IS A TEST THIS IS A TEST\n")
  end
  local time = stopStopWatch("echo.time")
  echo("Resolved in: "..time.."\n")
end

function addImage()
  setConsoleBackgroundImage(":/Mudlet_splashscreen_development.png")
end 
```

To perform the test first start the server and then connect Mudlet to localhost 10000.

Type ```lua bench()``` in Mudlet.
For the echo test type ```lua echoTest()``` 
To add a background image just type ```lua addImage()``` and then retest with image.

vadi2 commented:
Great job, speed wise runs well on my machine :)

vadi2 commented on src/TConsole.cpp:
```diff
@@ -288,7 +293,8 @@ TConsole::TConsole(Host* pH, ConsoleType type, QWidget* parent)
     splitter->setOrientation(Qt::Vertical);
     splitter->setHandleWidth(3);
-    splitter->setPalette(splitterPalette);
+    auto style_sheet = QStringLiteral("background-color: rgba(0,0,0,0)");
```
Could this and other cases be camelCase? That's the standard we follow on the C++ side.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -11991,66 +12027,101 @@ int TLuaInterpreter::setFgColor(lua_State* L)
+    QString windowName;
+    int r, g, b, alpha;
+
+    auto validRange = [](int number) { return number >= 0 and number <= 255; };
```
Maybe it's more readable with the and [synonym](https://en.wikipedia.org/wiki/Operators_in_C_and_C%2B%2B#C.2B.2B_operator_synonyms), no strong opinion, but we've used `&&` everywhere else :)

vadi2 reviewed (commented):
Works great, but there's a problem that the demo shows us - image stretching. What do you think allowing a selection of stretching/tiling/etc modes to help GUI makers?

Edru2 committed:
resetBgImage BgImageModes and code suggestions

Edru2 committed:
Merge branch 'development' into console-backgroundimage

Edru2 committed:
change order for :lower as it caused strange issues

Edru2 committed:
Geyser MiniConsole set/resetBackgroundImage update

Edru2 commented:
Added the mode parameter to give different modes like 1 "border", 2 "center", 3 "tile", 4 "style" with the style parameter it is even possible to add the background-image stylesheet. 
Also added ```resetConsoleBackgroundImage``` as I noticed that it was a bad idea to let setting the background-color overwrite the image.
(Updated the first post)
Ready for review :smiley:

breakone9r commented:
I just grabbed the linux appimage on my slow crappy laptop, and I can see no performance issues arising here. Even with a fairly complicated background image, I was moving thru many multiple rooms with no increased delays.

Compared to either the release 4.9.1 or the latest development branch.  I feel pretty confident in claiming there is no performance hit here.

demonnic reviewed (approved):
Works well for me, and the benchmarks look good. Also:
![image](https://user-images.githubusercontent.com/3660/93635131-c953bf00-f9bf-11ea-921c-e0aaacd7b5dc.png)

wiploo commented:
Tested and working. One another great PR from @Edru2!

vadi2 commented:
Pretty amazing work, thank you!
